src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.TryCloneFile.OSX.cs

```diff
@@ -0,0 +1,81 @@
+    {
+        private static bool TryCloneFile(string sourceFullPath, string destFullPath, bool overwrite)
+        {
+            // This helper function calls out to clonefile, and returns the error.
+            static bool TryCloneFile(string sourceFullPath, string destFullPath, int flags, out Interop.Error error)
+            {
+                if (Interop.@libc.clonefile(sourceFullPath, destFullPath, flags) == 0)
+                {
+                    // Success.
+                    error = Interop.Error.SUCCESS;
+                    return true;
+                }
+                error = Interop.Sys.GetLastError();
+                return false;
+            }
+
+            // Try to clone the file immediately, this will only succeed if the
+            // destination doesn't exist, so we don't worry about locking for this one.
+            int flags = Interop.@libc.CLONE_ACL;
+            Interop.Error error;
+            if (TryCloneFile(sourceFullPath, destFullPath, flags, out error)) return true;
+
+            // Some filesystems don't support ACLs, so may fail due to trying to copy ACLs.
+            // This will disable them and allow trying again (a maximum of 1 time).
+            if (error == Interop.Error.EINVAL)
+            {
+                flags = 0;
+                if (TryCloneFile(sourceFullPath, destFullPath, flags, out error)) return true;
+            }
+
+            // Try to delete the destination file if we're overwriting.
+            if (error == Interop.Error.EEXIST && overwrite)
+            {
+                // Delete the destination. This should fail on directories. Get a lock to the dest file to ensure we don't copy onto it when
+                // it's locked by something else, and then delete it. It should also fail if destination == source since it's already locked.
+                try
+                {
+                    using SafeFileHandle? dstHandle = SafeFileHandle.Open(destFullPath, FileMode.Open, FileAccess.ReadWrite,
+                        FileShare.None, FileOptions.None, preallocationSize: 0, createOpenException: CreateOpenExceptionForCopyFile);
+                    if (Interop.Sys.Unlink(destFullPath) < 0)
+                    {
+                        Interop.Error errorInfo = Interop.Sys.GetLastError();
+                        if (error != Interop.Error.ENOENT)
+                        {
+                            // Fall back to standard copy as an unexpected error has occurred.
+                            return false;
+                        }
+                    }
+                }
+                catch (FileNotFoundException)
+                {
+                    // We don't want to throw if it's just the file not existing, since we're trying to delete it.
+                }
+
+                // Try clonefile now we've deleted the destination file.
+                if (TryCloneFile(sourceFullPath, destFullPath, flags, out error)) return true;
```

Comment:
> also, how is this "2 weeks ago" lol.

You click the "Start review" button, start commenting and then return to it 2 weeks later to find out that you forgot to submit it (or just didn't finish it) ;-)

Reply:
> This is not ideal since we want to use CLONE_ACL if possible, we don't want to give up on that just because we had to delete the destination file.

I'm not following: `flags` was just set to 0 on the prior line... my style suggestion isn't changing anything about whether CLONE_ACL is used or not.

> also, how is this "2 weeks ago" lol.

I left the comments back then but apparently neglected to hit the submit button.

Reply:
It's only set to `0` in this if block, otherwise it would have `CLONE_ACL` as its value:
<img width="612" alt="image" src="https://github.com/dotnet/runtime/assets/25316035/3162393c-e182-4094-a71a-0a3c62db39ba">

Reply:
Ah, my comment was actually here:
https://github.com/dotnet/runtime/pull/79243/commits/f67bd18b5237853b3b22f9ea7a74fe524240c75a#diff-08d6878f7d95337f324dc91fea86beae0c4f60b72fcc6a61c14bb30fbc9a613bL72
but it appears the code changed between the time that I left the comment and submitted it.